xm-commons-cache/src/main/java/com/icthh/xm/commons/cache/service/DynamicCaffeineCacheManager.java

```diff
@@ -0,0 +1,140 @@
+import com.icthh.xm.commons.cache.TenantCacheManager;
+import com.icthh.xm.commons.cache.config.XmCacheConfig;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.tuple.Pair;
+import org.springframework.cache.caffeine.CaffeineCacheManager;
+import org.springframework.context.ApplicationListener;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import static com.icthh.xm.commons.cache.config.XmCacheConfig.CACHE_DEFAULTS;
+
+@RequiredArgsConstructor
+@Slf4j
+public class DynamicCaffeineCacheManager extends CaffeineCacheManager implements ApplicationListener<InitCachesEvent> {
+
+    private final ConcurrentMap<String, Supplier<Caffeine>> cacheCfgMap = new ConcurrentHashMap<>(16);
+    private final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
+
+    private final Ticker ticker;
+
+    @Override
+    public void onApplicationEvent(InitCachesEvent event) {
+        final String tenantName = event.getTenantKey().toUpperCase();
+
+        List<Pair<String, XmCacheConfig.XmCacheConfiguration>> cacheConfigurationList =
+            event.getCacheList().stream()
+                .map(cfg -> Pair.of(TenantCacheManager.buildKey(tenantName,cfg.getCacheName()), cfg))
+                .collect(Collectors.toList());
+
+        boolean isLocked = false;
+        try {
+            isLocked = lock.writeLock().tryLock(5, TimeUnit.SECONDS);
+            if (isLocked) {
+                final String cachePrefix = TenantCacheManager.buildPrefix(tenantName);
+
+                //remove all CFG for tenant
+                this.cacheCfgMap.entrySet().removeIf(entry -> entry.getKey().startsWith(cachePrefix));
+
+                //add suppliers for cache builders
+                for (Pair<String, XmCacheConfig.XmCacheConfiguration> pair: cacheConfigurationList) {
+                    this.cacheCfgMap.put(pair.getKey(), () -> buildSpec(pair.getValue()));
+                }
+            }
+        } catch (InterruptedException e) {
+            log.error("Tenant[" + tenantName + "] " +  e.getMessage(), e);
+        } finally {
+            if (isLocked) {
+                lock.writeLock().unlock();
+            }
+        }
+
+    }
+
+    @Override
+    public org.springframework.cache.Cache getCache(String name) {
+        if (!cacheCfgMap.keySet().contains(name)) {
+            throw new IllegalStateException("Cache with name " + name + " did not exist in cache mapping");
+        }
+        return super.getCache(name);
+    }
+
+    @Override
+    protected Cache<Object, Object> createNativeCaffeineCache(String name) {
+        boolean isLocked = false;
+        Cache<Object, Object> cache = null;
+        try {
```

Comment:
It seems the code that wraps logic execution with lock can be extracted to separate method that might avoid duplication.